File: Difficulty: Basic/Preorder Traversal/preorder-traversal.cpp
// NEW POTD Submission
/*
class Node
{
    int data;
    Node* left;
    Node* right;

    Node(int x){
        data = x;
        left = right = NULL;
    }
};
*/

class Solution {
  private:
    void traverse(Node* curr, vector<int>& result) {
        if (curr == nullptr) {
            return;
        }
        
       
        result.push_back(curr->data);
        
        
        traverse(curr->left, result);
        
        
        traverse(curr->right, result);
    }

  public:
    vector<int> preOrder(Node* root) {
        vector<int> result;
        traverse(root, result);
        return result;
    }
};

File: Difficulty: Basic/Preorder Traversal/preorder-traversal.py
class Solution:
    def preOrder(self, root):
        result = []

        def traverse(node):
            if not node:
                return
            result.append(node.data)   # Visit root
            traverse(node.left)        # Visit left subtree
            traverse(node.right)       # Visit right subtree

        traverse(root)
        return result